### ISSUE TITLE
ensure performance of tools on Dart 2 VM is on par with Dart 1

### ISSUE THREAD

**Issue by mraleph** [dart-lang/sdk#33257]
Currently known regressions:

- [ ] dart2js: around 2x regression in self compilation
- [ ] build_runner: 50% regression when building build_runner's own _test, 2x regression on some individual steps
- [ ] analyzer: indirectly confirmed via build_runner.

Most regressions seem to boil down to the cost of strong mode type checks: with parameter checks and implicit down-casts ignored hit on dart2js goes down to 10%.

**Comment by dgrove** [comment]
Any further updates here, @mraleph ?

**Comment by mraleph** [comment]
The first fix has landed on Friday. The work on another one is starting.

**Comment by dgrove** [comment]
Any updates here?

**Comment by mraleph** [comment]
No updates at the moment. Still working on this.

**Comment by handicraftsman** [comment]
Hmm?

**Comment by mraleph** [comment]
@handicraftsman please elaborate on the “hmm?” - leaving such comments is not really helpful

**Comment by handicraftsman** [comment]
Any news?

**Comment by mraleph** [comment]
Still working on this. Have another fix in the development that cuts another 50% of the regression.

**Comment by mraleph** [comment]
@handicraftsman is there any particular reason you are interested in this? does your application regress in Dart 2 mode? in this case it would be helpful if you shared some information - so that we could ensure that we address regressions in your applications as well.

**Comment by nex3** [comment]
Sass's speed regresses considerably in Dart 2 mode as well. Here's some data, collected using [the same techniques](https://github.com/sass/dart-sass/blob/master/perf.md) we use to collect cross-implementation performance data, tested against Dart 2.0.0-dev.62.0. Note that due to #28617, we have live users running against both script and application snapshots.

## Small Plain SCSS

Running on a file containing 4 instances of `.foo {a: b}`:

* Dart 1 mode with a script snapshot: 0.363s
* Dart 1 mode with an app snapshot: 0.117s
* Dart 2 mode with a script snapshot: 0.326s
* Dart 2 mode with an app snapshot: 0.172s

This is effectively just testing startup speed. It looks like Dart 2 actually improves startup speed with a script snapshot, but app snapshot startup speed is about 50% slower.

## Large Plain CSS

Running on a file containing 2^17 instances of `.foo {a: b}`:

* Dart 1 mode with a script snapshot: 2.479s
* Dart 1 mode with an app snapshot: 2.066s
* Dart 2 mode with a script snapshot: 2.893s
* Dart 2 mode with an app snapshot: 2.647s

This benchmark simulates a large-scale style system. We see about a 15% performance hit for script snapshots, up to about 30% for app snapshots. This is pretty concerning: one of the selling points of Dart Sass is its performance parity with the C++ implementation, but this will break that parity and thereby substantially hurt our value proposition.

**Comment by mraleph** [comment]
Thanks for trying this out @nex3 I will make sure to include Dart SASS in the set of tools I am benchmarking against.

**Comment by nex3** [comment]
FYI, it looks like the performance of 2.0.0-dev.64.1 is slow enough that it's causing [mass timeouts on Travis](https://travis-ci.org/sass/dart-sass/jobs/395153602) of tests that invoke Dart subprocesses.

**Comment by mraleph** [comment]
@nex3 are you sure it is performance related issue and not a legit timeout (e.g. due to sync-async semantics changes or something like that)?

**Comment by nex3** [comment]
Pretty sure; the tests pass on my fast desktop computer, but time out on Travis.

**Comment by dgrove** [comment]
@mraleph any updates on this?

**Comment by mraleph** [comment]
@dgrove the regression is down to 7-20% depending on where we measure. I expect some benefits from work on #33755 (most importantly using cid range machinery in JIT that was originally developed for AOT type testing stubs) - but I don't have timeline for that, @aartbik is just starting on that.

**Comment by nex3** [comment]
Here are the latest numbers for Dart Sass running on Dart 2.0.0-dev.67.0:

## Small Plain SCSS

Running on a file containing 4 instances of `.foo {a: b}`:

* Dart 1 mode with a script snapshot: 0.211s
* Dart 1 mode with an app snapshot: 0.066s
* Dart 2 mode with a script snapshot: 0.231s
* Dart 2 mode with an app snapshot: 0.108s

Dart 2 mode is about 10% slower than Dart 1 mode from a script snapshot, but 65% slower from an app snapshot. That said, both are substantially faster than 2.0.0-dev.62.0 in Dart 1 mode, so I'd count this as a win.

## Large Plain CSS

Running on a file containing 2^17 instances of `.foo {a: b}`:

* Dart 1 mode with a script snapshot: 1.964s
* Dart 1 mode with an app snapshot: 1.807s
* Dart 2 mode with a script snapshot: 2.005s
* Dart 2 mode with an app snapshot: 1.968s

Dart 2 mode is basically on par with Dart 2 mode from a script snapshot here, and about 10% slower from an app snapshot. Again, though, speed is overall substantially faster than 2.0.0-dev.62.0.

I'm satisfied with the speed of Dart 2 mode for deploying Dart Sass. However, startup time from source is still very slow--about 1.5s to run `dart bin/sass.dart --help`--which makes the development experience quite painful.

**Comment by rmacnak-google** [comment]
This is a duplicate of https://github.com/dart-lang/sdk/issues/28088, which was supposed to gate Dart 2.

**Comment by dgrove** [comment]
@mraleph Any updates on this?

**Comment by mraleph** [comment]
The progress can be tracked on https://github.com/dart-lang/sdk/issues/31798: we are landing changes, there is currently one CL in flight. With all changes combined I measure this on compiling large Dart application with dart2js:

- Before recent changes 400-410 seconds
- After all the changes 365-375 seconds
- Dart 1 mode: 350 seconds.

I have not measured other tools yet.

**Comment by dgrove** [comment]
@mraleph any updates here? Is anything else happening on this for 2.1?

**Comment by mraleph** [comment]
We had to disable one of the key optimizations we did. @mkustermann is working on reenabling it. We are going to return to optimizations after it is re-enabled.

**Comment by mraleph** [comment]
Move to 2.2 because the reland did not make it in time.

**Comment by nex3** [comment]
FYI, between this and the dropped support for `--no-preview-dart-2`, Dart Sass is blocked on upgrading its Dart version. We're currently pinned to Dart 2, which is the last stable version that has the performance we're looking for.

**Comment by mkustermann** [comment]
After waiting for more than a week for our buildbot to stabilize (and we still have flaky crashes left), we decided to land the bugfix now (06f9a9e35427d505f52893134a1e42eb166c49d2), which also re-enables the optimization.

**Comment by dgrove** [comment]
Can this be closed, @mraleph and @mkustermann ?

**Comment by dgrove** [comment]
Can this be closed, @mraleph and @mkustermann ?

**Comment by mraleph** [comment]
As far as I know we are still slower than Dart 1 sometimes, but we are currently not actively working on anything to reduce the overhead further. 

Need to gather the numbers again. 

@sstrickl could you take a stub at measuring where we stand currently with default mode versus `--experimental_unsafe_mode_use_at_your_own_risk`?

**Comment by mit-mit** [comment]
@sstrickl can we can an update on where we currently stand?

**Comment by sstrickl** [comment]
Just coming back to this after finishing up the RegExp work I was doing, so I'll try and have some initial numbers in here later this week. My initial benchmarking plans went awry when dart2js crashed while compiling the large code bases I was going to test it on, so I need to pick some alternative codebases for benchmarking. (That, and actually file an issue re: the crashes--I did let a member of the team know about them, but I should still create the issue as well for tracking.)

**Comment by sstrickl** [comment]
All the numbers below are using commit 0b40a46ab, since that's the last commit that could compile the codebases I've been using until an appropriate fix for issue #36516 is in place. However, that's only ~a month old (Mar 4), so the numbers shouldn't be too different on master.

-----

Here's the numbers from running the first handful of N's Sass benchmarks 10 times each, min and max included with the mean to give you an idea of the result range:

Benchmark        | Snapshot | Checked | Minimum | Mean   | Maximum
-----------------|----------|---------|---------|--------|--------
Small Plain      | Script   | Yes     | 0.18    | 0.196  | 0.21
"                |          | No      | 0.18    | 0.190  | 0.21
"                | App      | Yes     | 0.08    | 0.093  | 0.11
"                |          | No      | 0.08    | 0.091  | 0.11
Large Plain      | Script   | Yes     | 1.67    | 1.726  | 1.80
"                |          | No      | 1.64    | 1.691  | 1.76
"                | App      | Yes     | 1.67    | 1.737  | 1.79
"                |          | No      | 1.61    | 1.675  | 1.74
Preceding Sparse | Script   | Yes     | 1.81    | 1.866  | 1.98
"                |          | No      | 1.74    | 1.783  | 1.83
"                | App      | Yes     | 1.80    | 1.849  | 1.93
"                |          | No      | 1.74    | 1.792  | 1.87
Following Sparse | Script   | Yes     | 1.77    | 1.829  | 1.87
"                |          | No      | 1.69    | 1.746  | 1.85
"                | App      | Yes     | 1.78    | 1.827  | 1.92
"                |          | No      | 1.68    | 1.786  | 1.86
Preceding Dense  | Script   | Yes     | 2.66    | 2.717  | 2.79
"                |          | No      | 2.49    | 2.555  | 2.63
"                | App      | Yes     | 2.65    | 2.730  | 2.85
"                |          | No      | 2.50    | 2.597  | 2.71
Following Dense  | Script   | Yes     | 2.70    | 2.792  | 2.88
"                |          | No      | 2.60    | 2.664  | 2.81
"                | App      | Yes     | 2.74    | 2.851  | 2.90
"                |          | No      | 2.62    | 2.719  | 2.82

Here, the overhead of the strong type checks is ~5% on average, with the ranges on the checked vs. unchecked runs overlapping in almost all cases.

-----

For the two large codebases I tried compiling with dart2js (first compiling dart2js to a kernel snapshot and using that for the benchmark times, each codebase compiled five times each):

Benchmark        | Checked | Minimum | Mean   | Maximum
-----------------|---------|---------|--------|--------
Codebase 1       | Yes     | 374.74  | 381.24 | 389.62
"                | No      | 369.80  | 375.96 | 380.14
Codebase 2       | Yes     | 257.67  | 259.28 | 263.12
"                | No      | 252.28  | 256.96 | 260.94

Here, there's no real discernable overhead (~1%) to the runtime strong type checks, the variance on the runs dwarfs that.

-----

So while the dynamic checks do add a small bit of overhead, it seems much more reasonable than in prior testing.

**Comment by Stargator** [comment]
@sstrickl How often are these kinds of performance tests done? Before each release or only when certain components are changed?

**Comment by mraleph** [comment]
@Stargator we have a suite of performance tests which is automatically run for each commit to SDK repo in various configurations (native JIT/AOT, dart2js on V8 etc) on different hardware (different Intel variants, ARM32 and ARM64 hardware, mobile phones).

Due to technical complexities involved we don't track performance of some more real world scenarios there. However there is an expectation that performance on real world benchmarks is a function of performance on the benchmarks we are tracking. 

In general I would recommend anybody using Dart to also have at least some performance tracking on their side - so that you can catch potential regressions between releases and report them back to us for fixing - if we miss something.
